src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Updates project system service with new project info.
+/// </summary>
+/// <remarks>
+/// Used to figure out if the project system data is being added, updated or deleted
+/// and acts accordingly to update project system service.
+/// </remarks>
+internal partial class ProjectConfigurationStateManager : IDisposable
+{
+    private readonly ProjectSnapshotManagerDispatcher _projectSnapshotManagerDispatcher;
+    private readonly IRazorProjectService _projectService;
+    private readonly IProjectSnapshotManager _projectManager;
+    private readonly ILogger _logger;
+
+    private readonly AsyncBatchingWorkQueue<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> _workQueue;
+    private readonly CancellationTokenSource _disposalTokenSource;
+
+    public ProjectConfigurationStateManager(
+        ProjectSnapshotManagerDispatcher projectSnapshotManagerDispatcher,
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager)
+    {
+        _projectSnapshotManagerDispatcher = projectSnapshotManagerDispatcher;
+        _projectService = projectService;
+        _projectManager = projectManager;
+        _logger = loggerFactory.GetOrCreateLogger<ProjectConfigurationStateManager>();
+
+        _disposalTokenSource = new ();
+        _workQueue = new(
+            EnqueueDelay,
+            ProcessBatchAsync,
+            _disposalTokenSource.Token);
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            UpdateProject(workItem.ProjectKey, workItem.ProjectInfo);
+        }
+
+        return default;
+    }
+
+    internal TimeSpan EnqueueDelay { get; set; } = TimeSpan.FromMilliseconds(250);
```

Comment:
Note that tests can't actually change this, since it is queried in the constructor. To address this problem, I've been creating an extra constructor for tests that takes the delay. Here's an example:

https://github.com/dotnet/razor/blob/07e194a65a1a9f85fec3af041aace521300381b7/src/Razor/src/Microsoft.VisualStudio.Editor.Razor/DynamicFiles/BackgroundDocumentGenerator.cs#L32-L56